File: RuleEngine.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.longyun.flink;

import com.alibaba.fastjson.JSONObject;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer010;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.TableEnvironment;
import org.apache.flink.table.api.java.StreamTableEnvironment;

import java.util.Properties;
import java.util.Set;

/**
 * Skeleton for a Flink Streaming Job.
 *
 * <p>For a tutorial how to write a Flink streaming application, check the
 * tutorials and examples on the <a href="http://flink.apache.org/docs/stable/">Flink Website</a>.
 *
 * <p>To package your application into a JAR file for execution, run
 * 'mvn clean package' on the command line.
 *
 * <p>If you change the name of the main class (with the public static void main(String[] args))
 * method, change the respective entry in the POM.xml file (simply search for 'mainClass').
 */
public class RuleEngine {

    final static String brokers = "res-spark-0001:9092,res-spark-0002:9092,res-spark-0003:9092";

    /**
     *
     * @param args
     * @throws Exception
     */
	public static void main(String[] args) throws Exception {
	   //1. set kafka consumer properties
        Properties properties = new Properties();
        properties.setProperty("bootstrap.servers", brokers);
        properties.setProperty("group.id", "flink");
        properties.setProperty("session.timeout.ms", "30000");
        properties.setProperty("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        properties.setProperty("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");

        //Partition discovery
        properties.setProperty("flink.partition-discovery.interval-millis", "30000");

		// set up the streaming execution environment
		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //启用checkpoint
        env.enableCheckpointing(5000);

        DataStream<String> stream = env.addSource(new FlinkKafkaConsumer010<String>("iot-src", new SimpleStringSchema(), properties));

        String template = "{\"state\": {\"reported\": {\"temperature\": 73.09}},\"clientToken\": \"client-27\",\"timeStamp\": 1533709399480}";

        JSONObject json = JSONObject.parseObject(template);
        Set<String>  keys = json.keySet();
        String[] fieldNames = new String[keys.size()];
        keys.toArray(fieldNames);

        // get StreamTableEnvironment
        // registration of a DataSet in a BatchTableEnvironment is equivalent
        StreamTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);


        // register the DataStream as Table "myTable" with fields
        tableEnv.registerDataStream("sensor", stream);

        Table table = tableEnv.sqlQuery("select * from sensor");
//        Table table = tableEnv.fromDataStream(stream);
//        TableSchema schema = table.getSchema();
        System.out.println("--------------------table schema--------------------");
        table.printSchema();



		// execute program
		env.execute("Flink Rule Engine");
	}
}

File: rule-engine-flink-table/src/main/scala/com/inspur/cloud/res/RuleEngine.scala
package com.inspur.cloud.res

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.core.fs.FileSystem.WriteMode
import org.apache.flink.streaming.api.scala._
import org.apache.flink.table.api.{TableEnvironment, Types}
import org.apache.flink.table.descriptors.{Csv, Json, Kafka, Schema}
import org.apache.flink.table.sinks.{CsvTableSink, PrintTableSink}

import scala.collection.mutable.ArrayBuffer

/**
  * Skeleton for a Flink Streaming Job.
  *
  * For a tutorial how to write a Flink streaming application, check the
  * tutorials and examples on the <a href="http://flink.apache.org/docs/stable/">Flink Website</a>.
  *
  * To package your application into a JAR file for execution, run
  * 'mvn clean package' on the command line.
  *
  * If you change the name of the main class (with the public static void main(String[] args))
  * method, change the respective entry in the POM.xml file (simply search for 'mainClass').
  */
object RuleEngine {

  val brokers = "res-spark-0001:9092,res-spark-0002:9092,res-spark-0003:9092"

  val sourceTopic = "iot-src"

  /**
    *
    * @param args
    */
  def main(args: Array[String]) {

    // set up the streaming execution environment
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    //启用checkpoint
//    env.enableCheckpointing(5000)

//    val template: String = "{\"state\": \"online\",\"clienttoken\": \"client-27\",\"timestamp\": 1533709399480}"
    val tableEnv = TableEnvironment.getTableEnvironment(env)

    val descriptor = new Kafka().version("0.11")
        .topic(sourceTopic)
        .startFromEarliest()

//        .startFromLatest()
        .property("bootstrap.servers", brokers)
        .property("group.id", "res")
        .property("session.timeout.ms", "30000")
        .sinkPartitionerFixed()

    val tableDescriptor = tableEnv.connect(descriptor)
      .withFormat(new Json()
        .failOnMissingField(false)
        .deriveSchema())
      .withSchema(new Schema()
        .field("clienttoken", Types.STRING)
        .field("timestamp", Types.LONG)
        .field("state", Types.ROW(Array[String]("status", "value"), Array[TypeInformation[_]](Types.STRING, Types.INT)))
//          .field("state", Types.STRING)
      )
      .inAppendMode().registerTableSource("sensor")
    val table = tableEnv.sqlQuery("select clienttoken, `timestamp`, state from sensor where `value` is not null")

    println("--------------------table sensor schema--------------------")
    table.printSchema()


    tableEnv.registerTableSink("console",
      Array[String]("f0", "f1", "f2"),
      Array[TypeInformation[_]](Types.STRING,Types.LONG, Types.ROW(Types.STRING, Types.INT)),
      new PrintTableSink
    )

    table.insertInto("console")

    // execute program
    env.execute("Flink Rule Engine Application")
  }
}
